### ISSUE TITLE
Incorrect encoding for the names of music files

### ISSUE THREAD

**Issue by anfimovir** [telegramdesktop/tdesktop#1513]
Telegram Desktop 0.9.18

Screenshot: 1 - from telegram interface.
Screenshot: 2 - from OS.

1.
![1](https://cloud.githubusercontent.com/assets/10336636/12269317/e268af58-b961-11e5-8296-1abe538c0226.PNG)

2.
![2](https://cloud.githubusercontent.com/assets/10336636/12269318/e279a646-b961-11e5-86cf-93871532a578.PNG)

Please add support cyrillic letters in the file names.

**Comment by DaFri-Nochiterov** [comment]
I think TDesktop print not a filename. It's tag of your audio file. Did you try to change the tags?

**Comment by telegramdesktop** [comment]
@DaFri-Nochiterov The tags are read like utf-8 strings. If they are in some local encoding it will be bad :( didn't think of how to determine the encoding yet.

**Comment by Zenexer** [comment]
I've seen the same issue with UTF-8 tags.

**Comment by telegramdesktop** [comment]
@Zenexer UTF-8 tags should be correct, I need an example of wrong reading of UTF-8 tags.

**Comment by Zenexer** [comment]
If it happens again I'll be sure to save a copy.  I believe the files I saw
were from a Spotify bot that offered previews.

**Comment by formatq** [comment]
Hi. I have an example mp3 file with a bad read tags. [https://drive.google.com/file/d/0BwgJlWodegEDZTNJYnNTSjNhVkE/view](https://drive.google.com/file/d/0BwgJlWodegEDZTNJYnNTSjNhVkE/view) 

![How it looks like in tDesktop](http://webdav.aplana.com/monosnap/fm_160212_104328.png)

v.0.9.21dev

**Comment by DaFri-Nochiterov** [comment]
@formatq, try manually change tags with AIMP3 Tag Editor other same software. Not sure, but it can solve this problem.

**Comment by formatq** [comment]
@DaFri-Nochiterov thanks for the solution. But I think there is a way to decode tags in correct encoding on the fly.

**Comment by tdesktop-closer** [comment]
Hey there!

We're automatically closing this issue since there was no activity in this issue since 362 days ago. We therefore assume that the user has lost interest or resolved the problem on their own. Closed issues that remain inactive for a long period may get automatically locked.

Don't worry though; if this is in error, let us know with a comment and we'll be happy to reopen the issue.

Thanks!

<sub>(Please note that this is an [automated](https://github.com/auchri/github_auto_closer) comment.)</sub>

**Comment by mana71** [comment]
Hi!
I have the same issue with russian letters on telegram desktop, telegram android, telegram IOS. How can I resolve the problem with cyrillic in filenames in Telegram?

Thank you in advance)

**Comment by mana71** [comment]
![bad_encoding](https://user-images.githubusercontent.com/24596321/39993533-d8743540-577e-11e8-8545-eb0fe0951865.png)

**Comment by mana71** [comment]
![bad_encoding](https://user-images.githubusercontent.com/24596321/39995945-8fe87c70-5786-11e8-83e8-2be9a41ac30e.png)
You can see that in Windows explorer these tracks have correct encoding but in telegram (desktop or mobile) I see strange characters. I attached the track with incorrect characters in zip file.



[tracks.zip](https://github.com/telegramdesktop/tdesktop/files/2000530/tracks.zip)

**Comment by stek29** [comment]
@mana71 first of all, Telegram uses tags in file and not the filename to get artist and song names.

second: Here's how macOS displays that file:
![screen shot of finder preview with ""incorrect encoding""](https://user-images.githubusercontent.com/11808223/40434662-7bee388a-5eb7-11e8-9e74-86d6cda5820c.png)

---

third: Let's have a look at the file you've sent -- open it in hex editor (or just hexdump it):
```
❯ xxd Мистер\ Малой\ -\ дЕнЕг\ д0лЖен.mp3| head -n3
00000000: 4944 3303 0000 0000 0776 5449 5432 0000  ID3......vTIT2..
00000010: 000b 0000 00ca e0e9 f4ee ede0 e2f2 fb54  ...............T
00000020: 414c 4200 0000 1500 0000 c020 e820 c120  ALB........ . .
```

As you can see, it starts with ID3 -- that's a magic for ID3v2. Two following bytes indicate exact version: 2.3.0. [Here's the informal standard](http://id3.org/id3v2.3.0).
tldr on the format:
> As the tag consists of a tag header and a tag body with one or more frames, all the frames consists of a frame header followed by one or more fields containing the actual information. The layout of the frame header:


Here's what frames look like:
>     Frame ID       $xx xx xx xx (four characters)
>     Size           $xx xx xx xx
>     Flags          $xx xx

So, for the song title (TIT2) frame:
>     Frame ID       $54 49 54 32 (TIT2)
>     Size           $00 00 00 0B
>     Flags          $00 00
>     Actual data    $00 CA E0 E9 F4 EE ED E0 E2 F2 FB

So, let's see how to parse Data in the Text Information Frames (which TIT2 is):

>     <Header for 'Text information frame', ID: "T000" - "TZZZ", excluding "TXXX" described in 4.2.2.>
>     Text encoding    $xx
>     Information    <text string according to encoding>

Ok, so now we have:
>     Frame ID       $54 49 54 32 (TIT2)
>     Size           $00 00 00 0B
>     Flags          $00 00
>     Text encoding  $00
>     Information    $CA E0 E9 F4 EE ED E0 E2 F2 FB

So, what is $00?

>  Frames that allow different types of text encoding have a text encoding description byte directly after the frame size. If ISO-8859-1 is used this byte should be $00, if Unicode is used it should be $01. Strings dependent on encoding is represented as <text string according to encoding>, or <full text string according to encoding> if newlines are allowed. Any empty Unicode strings which are NULL-terminated may have the Unicode BOM followed by a Unicode NULL ($FF FE 00 00 or $FE FF 00 00).

aha, so `$CA E0 E9 F4 EE ED E0 E2 F2 FB` is `ISO-8859-1` string!

Now, let's decode it:

```
In [1]: from binascii import unhexlify

In [2]: unhexlify('CAE0E9F4EEEDE0E2F2FB').decode('ISO-8859-1')
Out[2]: 'Êàéôîíàâòû'
```

That's exactly what macOS and TDesktop show.

Now, bonus step: what if we decode it as [CP1251](https://en.wikipedia.org/wiki/Windows-1251) -- encoding that Windows loves to use for Cyrillic languages?

```
In [3]: unhexlify('CAE0E9F4EEEDE0E2F2FB').decode('CP1251')
Out[3]: 'Кайфонавты'
```

Well, that looks like a valid name. And yet another reason to hate how windows deals with text, completely ignoring standards and not even thinking about people who use different encodings. Another format where this causes a lot of issues is ZIP archive, btw.

Feel free to comment if you find a file conforming to standards but being parsed incorrectly by TDesktop.

**Comment by stek29** [comment]
@mana71 and I just realized that even on your screenshot windows explorer shows those broken names.
